### PR TITLE
[CS] Unify ReturnStmt handling

### PR DESCRIPTION
**Pull request by hamishknight**
Unify the implementation between single-expression and multi-statement closures. Because we're now storing a contextual type for single expression closure returns, update a code completion test to bring its behavior inline with the multi-statement case.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSSyntacticElement.cpp [hunk 1]
```diff
@@ -1281,9 +1255,18 @@ class SyntacticElementConstraintGenerator
     }
 
     auto contextualResultInfo = getContextualResultInfo();
+
+    // Single-expression closures need their contextual type locator anchored
+    // on the closure itself.
+    if (context.isSingleExpressionClosure(cs) && returnStmt->hasResult()) {
```
Comment (xedin) [inline lib/Sema/CSSyntacticElement.cpp:1261]: Can this be a part of `getContextualResultInfo()` so we don't have to mutate it post factum?

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/CSSyntacticElement.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/CSSyntacticElement.cpp]: Good idea! How's this?

**Comment by hamishknight** [comment]
Source compat run passed except for an issue that looks unrelated

File: lib/Sema/CSSyntacticElement.cpp [hunk 2]
```diff
@@ -1337,8 +1311,19 @@ class SyntacticElementConstraintGenerator
...
-            getAsExpr<ClosureExpr>(funcRef->getAbstractClosureExpr()))
-      return {cs.getClosureType(closure)->getResult(), CTP_ClosureResult};
+            getAsExpr<ClosureExpr>(funcRef->getAbstractClosureExpr())) {
+      // Single-expression closures need their contextual type locator anchored
+      // on the closure itself. Otherwise we use the default contextual type
+      // locator, which will be created for us.
+      ConstraintLocator *loc = nullptr;
+      if (context.isSingleExpressionClosure(cs) && returnStmt->hasResult()) {
+        loc = cs.getConstraintLocator(
+            context.getAsAbstractClosureExpr().get(),
```
Comment (xedin) [inline lib/Sema/CSSyntacticElement.cpp:1321]: Nit: you can use `closure` here instead of re-fetching it

**Review by xedin — APPROVED** [review]
Looks great!

File: lib/Sema/CSSyntacticElement.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/CSSyntacticElement.cpp]: Ack, good catch

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by ahoppen — COMMENTED** [review]
Change in code completion behavior looks reasonable to me.

**Pull request merged**